File: sdk/js/src/schemas/literal.ts
import type { ParseContext, SchemaNode } from "../types.js";
import { BaseSchema } from "./base.js";
import { ISSUE_CODES } from "../issue-codes.js";

type LiteralValue = string | number | boolean | null;

export class LiteralSchema<T extends LiteralValue> extends BaseSchema<T, T> {
  private _value: T;

  constructor(value: T) {
    super();
    this._value = value;
  }

  _validate(input: unknown, ctx: ParseContext): unknown {
    if (input !== this._value) {
      ctx.issues.push({
        code: ISSUE_CODES.INVALID_LITERAL,
        message: `Expected literal ${String(this._value)}, received ${String(input)}`,
        path: [...ctx.path],
        expected: String(this._value),
        received: String(input),
      });
      return undefined;
    }
    return input;
  }

  _toNode(): SchemaNode {
    const node = { kind: "literal" as const, value: this._value };
    this._addDefault(node as unknown as SchemaNode);
    return node as unknown as SchemaNode;
  }
}

File: sdk/rust/src/schemas/literal.rs
use serde_json::{json, Value};

use crate::issue_codes::*;
use crate::schema::{
    ParseContext, Schema, DescribeOpts, add_metadata_to_node, build_describe_metadata,
    merge_metadata, reserved_metadata_keys, validate_metadata_keys,
};
use crate::types::{PathSegment, ValidationIssue};

/// Schema that matches a single literal value.
#[derive(Debug, Clone)]
pub struct LiteralSchema {
    pub value: Value,
    pub metadata: Option<Value>,
}

impl LiteralSchema {
    pub fn new(value: Value) -> Self {
        LiteralSchema {
            value,
            metadata: None,
        }
    }

    pub fn describe(mut self, description: &str, opts: Option<&DescribeOpts>) -> Self {
        let new_meta = build_describe_metadata(description, opts);
        match &mut self.metadata {
            Some(existing) => merge_metadata(existing, &new_meta),
            None => self.metadata = Some(new_meta),
        }
        self
    }

    pub fn with_metadata(mut self, meta: Value, replace: bool) -> Self {
        validate_metadata_keys(&meta);
        if replace {
            let reserved = reserved_metadata_keys();
            let mut preserved = serde_json::Map::new();
            if let Some(Value::Object(existing)) = &self.metadata {
                for (k, v) in existing {
                    if reserved.contains(k.as_str()) {
                        preserved.insert(k.clone(), v.clone());
                    }
                }
            }
            if let Value::Object(new_map) = &meta {
                for (k, v) in new_map {
                    preserved.insert(k.clone(), v.clone());
                }
            }
            self.metadata = Some(Value::Object(preserved));
        } else {
            match &mut self.metadata {
                Some(existing) => merge_metadata(existing, &meta),
                None => self.metadata = Some(meta),
            }
        }
        self
    }
}

fn display_value(v: &Value) -> String {
    match v {
        Value::String(s) => s.clone(),
        Value::Number(n) => n.to_string(),
        Value::Bool(b) => b.to_string(),
        Value::Null => "null".to_string(),
        other => other.to_string(),
    }
}

impl Schema for LiteralSchema {
    fn kind(&self) -> &str {
        "literal"
    }

    fn parse_value(
        &self,
        input: &Value,
        path: &[PathSegment],
        _ctx: &ParseContext,
    ) -> Result<Value, Vec<ValidationIssue>> {
        if input == &self.value {
            Ok(input.clone())
        } else {
            Err(vec![ValidationIssue {
                code: INVALID_LITERAL.to_string(),
                path: path.to_vec(),
                expected: display_value(&self.value),
                received: display_value(input),
                meta: None,
            }])
        }
    }

    fn export_node(&self) -> Value {
        let mut node = json!({"kind": "literal", "value": self.value});
        add_metadata_to_node(&mut node, &self.metadata);
        node
    }
}
